### PR TITLE
[WIP] feature(repo): accept multi-store configuration

### PR DESCRIPTION
**Pull request by liqueflies**
Hello!

after some work I am going to submit you this first review for sane-shopify multistore!

As you said in #147, the most important thing where keep the store with already synched stores and migrate.
Also sync storeName indie document.

I came up with this solution in that way:

- secrets was already a document, so I ended up removing the unique key generating a new one for each store.
- The idea for the achievement was keeping the Provider component the same, targeting only 1 store; Inside the ShopifyTool component I fetch each key on the secrets table and put inside the provider the storeName that will populate the Provider.
- I fixed also the problem (that I noticed but I thought was removed feature) of Progress component.
- I added a Badge component in order to show into the document the store where product / collection is coming from
- I added store name as a description (until is archived)
- I check when import document before archiving them, if are coming from the same store: with this check we are avoiding to delete document that are not present in the store when we are syncing
- I made a fresh design both for the TabTool and then the ComponentInput inside the document
- Everything is made using @sanity/ui

What's missing?

basically webhook. I am going to ask your check for the rest and help me with that.

I leave here some screenshot but of course I suggest you to download my branch linking into a project of yours :)

Hope this can be a good news to you 🎉 

Screenshot 2021-03-25 at 10 21 13

Screenshot 2021-03-25 at 10 20 39

Screenshot 2021-03-25 at 10 20 59

Screenshot 2021-03-24 at 18 39 03

### CODE REVIEW

**Comment by liqueflies** [comment]
UPDATE: I saw you made a fix in syncState that I also made but your way was better, so I rebased with your main

**Comment by liqueflies** [comment]
UPDATE 2: checking now for webhook you are going to setup your implementation of webhook so you are providing already the correct secrets for shopify

File: packages/sanity-plugin/src/ShopifyTool/Setup.tsx [hunk 1]
```diff
@@ -1,36 +1,27 @@
 import { uniqueId } from 'lodash'
 import * as React from 'react'
+import { Stack, Button, Box, Text, TextInput } from '@sanity/ui'
```
Comment (good-idea) [inline packages/sanity-plugin/src/ShopifyTool/Setup.tsx:3]: 👏👏👏

File: packages/sanity-plugin/src/sanityDocuments/collection.tsx [hunk 2]
```diff
@@ -94,7 +94,7 @@ export const createCollectionDocument = ({
         const itemTitle = sourceData ? title || sourceData.title : title
         const alt = `Image for ${itemTitle}`
         const src = sourceData?.image?.w100
-        const subtitle = archived ? '📁 Archived' : undefined
+        const subtitle = archived ? '📁 Archived' : sourceData?.shopName || ''
```
Comment (good-idea) [inline packages/sanity-plugin/src/sanityDocuments/collection.tsx:97]: Could we only include the shopName in the subtitle if the user has multiple storefronts added?

File: packages/sync-utils/src/sanity/syncSanityDocument.ts [hunk 3]
```diff
@@ -149,7 +151,8 @@ export const createSyncSanityDocument = (
   const syncItem = async (
     item: Product | Collection
   ): Promise<SyncOperation> => {
-    const docInfo = prepareDocument(item)
+    const { shopName } = shopifyClient
+    const docInfo = prepareDocument({ ...item, shopName })
     const existingDoc = await getSanityDocByShopifyId(item.id)
 
     /* If the document exists and is up to date, skip */
```
Comment (good-idea) [inline packages/sync-utils/src/sanity/syncSanityDocument.ts:160]: could you add `shopName` to the list of keys below? This will make sure all existing documents get updated when running a new sync.

File: packages/sync-utils/src/sanity/syncSanityDocument.ts [hunk 4]
```diff
@@ -115,7 +116,8 @@ const mergeExistingFields = (
 
 export const createSyncSanityDocument = (
   client: SanityClient,
-  cache: SanityCache
+  cache: SanityCache,
+  shopifyClient: ShopifyClient
```
Comment (good-idea) [inline packages/sync-utils/src/sanity/syncSanityDocument.ts]: I'd like to avoid using any Shopify client things within `/src/sanity`... There is no issue with what you're doing here - but I want to discourage others from doing any shopify fetching from within these functions.

How about we create a `SaneShopifyContext` interface that (for now) is just:

```
interface SaneShopifyContext {
  shopName: string
}
```

and pass that in instead?

File: packages/sync-utils/src/shopify/fetchItemById.ts [hunk 5]
```diff
@@ -7,6 +7,9 @@ import { fetchAllProductCollections } from './fetchShopifyProduct'
 
 export const NODE_QUERY = gql`
   query NodeQuery($id: ID!) {
+    # shop {
```
Comment (good-idea) [inline packages/sync-utils/src/shopify/fetchItemById.ts]: Let's remove these lines if we aren't using them

File: packages/sync-utils/src/shopify/fetchShopifyProduct.ts [hunk 6]
```diff
@@ -59,6 +59,9 @@ const PRODUCT_BY_ID = gql`
     $collectionsAfter: String
   ) {
     node(id: $id) {
+      # shop {
```
Comment (good-idea) [inline packages/sync-utils/src/shopify/fetchShopifyProduct.ts]: same as above

**Review by good-idea — CHANGES_REQUESTED** [review]
THANK YOU for all of this great work! I'm excited to get this merged. I'll try it out on some of my current projects.

**Comment by good-idea** [comment]
I think it would also be good to add a disclaimer about adding multiple storefronts - and provide a link to a section of the docs here about caveats when working with multiple storefronts. Essentially, that they will need to handle their own logic for multiple carts.

I'm imagining a scenario where I want to create a "popup" site, selling products made by my friends - where all products are displayed on the website. Just thinking out loud here..

File: packages/sanity-plugin/src/ShopifyTool/Setup.tsx (hunk 1 above)
Comment (liqueflies) [reply packages/sanity-plugin/src/ShopifyTool/Setup.tsx:3]: Can't understand what you mean :(
What I have to do there?

File: packages/sanity-plugin/src/sanityDocuments/collection.tsx (hunk 2 above)
Comment (liqueflies) [reply packages/sanity-plugin/src/sanityDocuments/collection.tsx:97]: If product is archived it probably do not have a "store" anymore (deleted, archived or so)
In my opinion is better to keep the "Archived" label, what do you think?

Comment (liqueflies) [reply packages/sanity-plugin/src/sanityDocuments/collection.tsx:97]: Ah ok get it, you don't want to include if there is only "one" store connected.
Is not easy because we should make a count query to the secrets document... any better ideas?

File: packages/sync-utils/src/sanity/syncSanityDocument.ts (hunk 3 above)
Comment (liqueflies) [reply packages/sync-utils/src/sanity/syncSanityDocument.ts:160]: Ok let's check if it's better now

File: packages/sync-utils/src/sanity/syncSanityDocument.ts (hunk 4 above)
Comment (liqueflies) [reply packages/sync-utils/src/sanity/syncSanityDocument.ts]: Ok let me know if it's better not

**Comment by liqueflies** [comment]
> I think it would also be good to add a disclaimer about adding multiple storefronts - and provide a link to a section of the docs here about caveats when working with multiple storefronts. Essentially, that they will need to handle their own logic for multiple carts.
> 
> I'm imagining a scenario where I want to create a "popup" site, selling products made by my friends - where all products are displayed on the website. Just thinking out loud here..

In my particular case I was able to handle multiple listing based on location (https://www.shopify.com/plus/solutions/international-ecommerce)

So what've done is having all countries in a singular sanity instance.
I've made a document called region (US, EU, ROW) for example.

Example:

on document fields add:
```
  {
    type: 'reference',
    name: 'region',
    title: 'Display in',
    // hidden: true,
    to: [{ type: 'region' }]
  }
```

Than in desk structure I filter document from references on that region (https://www.sanity.io/docs/dynamically-group-list-items-with-a-groq-filter)

Talking about storefront I have to came up with different websites for each region:
`eu.mystore.com us.mystore.com row.mystore.com`

I came up using the same codebase for each website with an ENV variable region (EU, US, ROW) and I make queries using the reference on documents and shopify storefront and token specific for each website.

So each website has it's storefront shopify keys and different cart, the code from frontend is the same!

**Comment by liqueflies** [comment]
Can I do something different for this?

**Comment by good-idea** [comment]
Added a couple things and published this in another branch because I didn't know I could push to forked repo branches!

Thank you for all of your work on this!
